### DISCUSSION TITLE
Will it work with MMM-GoogleCalendar module?

### DISCUSSION THREAD

**Discussion by parrty** [MMRIZE/MMM-CalendarExt3#5] [General]
Thanks for making this. It's very nice!   The MMM-GoogleCalendar module uses API output which is not ical format.  If this module is able to format the api output the same as the default calendar then that would be amazing. 

Thanks

**Comment by eouia** [comment]
I think it's possible because MMM-GoogleClanedar seems to emit `CALENDAR_EVENTS` notification. I'll test it.

**Comment by eouia** [comment]
Well, I tried to install MMM-googlecalendar (asbjorn/MMM-googlecalendar) but failed. I don't know why. Anyway, if the module can emit CALENDAR_EVENTS notification, CX3 can gather needed info from it. (It might need some manipulation with `eventTransformer`, but I cannot test it because my MMM-googlecalendar fails to load itself)

**Reply by CurlyQ12391** [reply]
Hello, what about with the randomBrainstormer/MMM-GoogleCalendar module? In looking at the module's config there is a "CALENDAR_EVENTS" notification; however, I am not sure what manipulation might be required for CX3's eventTransformer.

**Reply by eouia** [reply]
Could you show me the log of this result?

```js
eventTransformer: (e) => {
  console.log('event structure:', e)
  return e
}
```
To make things simple, only 1 simple calendar with very few events would be better to be used in your GoogleCalendar module.

You can see the result on your dev console. Let's see what is showing, then discuss what would be needed.

**Reply by eouia** [reply]
@CurlyQ12391, and this is just a simple question. What is the better point on the direct google calendar module than the default ical based calendar module? this is just pure curiosity.

**Reply by CurlyQ12391** [reply]
Hey, I apologize because I am very new to all this. How would I show you the log of that specific action? I can confirm that MMM-GoogleCalendar is running and loading events on my Magic Mirror. 

I use a shared family calendar for google and that type of calendar doesn't have an iCal ID so the default module wont work. I use MMM-GoogleCalendar in order to have the same basic view of the default calendar with the ability to pull the actual shared family events. Does that answer your last question?

**Reply by eouia** [reply]
yup. I think it was the reason. :)
Anyway I updated. Try it.

**Reply by CurlyQ12391** [reply]
I updated the module, thank you by the way for incorporating these changes so quickly!!

However, I am not seeing the events transfer over to the CX3 module. Would it be possible to share my config.js with you for troubleshooting? If so, would you prefer in this discussion or for me to open up a separate issue?

**Reply by eouia** [reply]
Whatever you want. Plz show me the config and screen capture(or photo taken by your phone) together to know things better

**Reply by eouia** [reply]
By the way, it’s too late time for me so I’ll check it tomorrow. Sorry.

**Reply by CurlyQ12391** [reply]
perfect, thank you! I believe I scrubbed any sensitive info from the config...also apologies I cannot seem to finagle the entire thing as "code" :|

`
let config = {
	address: "0.0.0.0",
	port: 8080,
	basePath: "/", 
	ipWhitelist: ["127.0.0.1", "::ffff:127.0.0.1", "::1"], 
format :
	useHttps: false, 		// Support HTTPS or not, default "false" will use HTTP
	httpsPrivateKey: "", 	// HTTPS private key path, only require when useHttps is true
	httpsCertificate: "", 	// HTTPS Certificate path, only require when useHttps is true

	language: "en",
	locale: "en-US",
	logLevel: ["INFO", "LOG", "WARN", "ERROR"], // Add "DEBUG" for even more logging
	timeFormat: 12,
	units: "imperial",
	// serverOnly:  true/false/"local" ,
	// local for armv6l processors, default
	//   starts serveronly and then starts chrome browser
	// false, default for all NON-armv6l devices
	// true, force serveronly mode, because you want to.. no UI on this device

	modules: [
		 {
        module: 'MMM-pages',
        config: {
                modules:
                    [
                    [ "MMM-GoogleDocs-Notes", "newsfeed", "MMM-CalendarExt3" ],
                    [ "MMM-RemoteCompliments" ],
                    [ "MMM-GoogleCalendar" ],
                    ],
                fixed: [ "clock", "alert", "updatenotification", "MMM-OpenWeatherForecast", "MMM-page-indicator","MMM-Cursor" ],
                hiddenPages: {
                    "": [ "" ],
							},
				}
		},
		{
        module: 'MMM-page-indicator',
        position: 'bottom_bar',
        config: {
            pages: 3,
        }
		},
		{
			module: "alert",
		},
		{
			module: "updatenotification",
			position: "top_bar"
		},
		{
			module: "clock",
			position: "top_left",
			defaults: {
				displayType: "digital", // options: digital, analog, both
				displaySeconds: true,
				showPeriod: true,
				showPeriodUpper: false,
				clockBold: false,
				showDate: true,
				showTime: true,
				showWeek: false,
				dateFormat: "dddd, LL",
				}
		},
		/* default/calendar module configuration */
		{
			module: 'MMM-GoogleCalendar',
			header: "Skaggs Family Calendar",
			position: "middle_center",
			config: {
				calendars: [
					{
					  symbol: "calendar-week",
					  calendarID: "familyxxxxx@group.calendar.google.com"
					},
					// add another calendar HERE if needed
				],
			}
		},
		{
			module: "MMM-CalendarExt3",
			position: "middle_center",
			title: "",
			  config: {
				mode: "week",
				weekIndex: -1, // Which week from this week starts in a view. Ignored on mode 'month' 
				weeksInView: 2, //  How many weeks will be displayed. Ignored on mode 'month'
				instanceId: null,
				firstDayOfWeek: 1, // 0: Sunday, 1: Monday
				minimalDaysOfNewYear: 4, // When the first week of new year starts in your country.
				locale: "en-US", 
				maxEventLines: 10,
				firstDayOfWeek: 1,
				useSymbol: true,
				calendarSet: null,
			}
		},
		{
			module: "newsfeed",
			position: "bottom_left",
			config: {
				feeds: [
					{
						title: "New York Times",
						url: "https://rss.nytimes.com/services/xml/rss/nyt/HomePage.xml"
					}
				],
				showSourceTitle: true,
				showPublishDate: true,
				broadcastNewsFeeds: true,
				broadcastNewsUpdates: true
			}
		},
		{
			module: 'MMM-GoogleDocs-Notes',
			header: "Grocery List",
			position: 'top_center',
			classes: 'default everyone',
			config: {
				maxNotes: 2,
				pollFrequency: 5 * 60 * 1000, //5 minutes;
				showDatePosted: false,
				showCheckbox: true,
				notesPrefix: 'MMM-GroceryList'
					}
		},
		{
			module: "MMM-OpenWeatherForecast",
			position: "top_right",
			header: "Current Conditions",
			config: {
			  apikey: "zzzzz, //only string here
			  latitude: "33.9159",            			  //number works here
			  longitude: "-112.136",			          //so does a string
			  units: "imperial",
			  iconset: "3c",
			  colored: true,
			  concise: false,
			  requestDelay: "2000",
			  showFeelsLikeTemp: false,
			  showSummary: false,

			  showCurrentConditions: true,			  
			  showExtraCurrentConditions: true,
			  extraCurrentConditions: {
				highLowTemp: true,
				precipitation: true,
				sunrise: false,
				sunset: false,
				wind: true,
				barometricPressure: false,
				humidity: false,
				dewPoint: false,
				uvIndex: false,
				visibility: false
			  },

			  forecastLayout: "tiled",
			  forecastHeaderText: "Forecast",

				showHourlyForecast: false,
				showHourlyTableHeaderRow: false,
				hourlyForecastTableHeaderText: "Hourly",
				hourlyForecastInterval: 3,
				maxHourliesToShow: 3,
				hourlyExtras: {
				  precipitation: false,
				  wind: false,
				  barometricPressure: false,
				  humidity: false,
				  dewPoint: false,
				  uvIndex: false,
				  visibility: false
				},

			  dailyForecastTableHeaderText: "Daily",
			  showDailyForecast: true,
			  showDailyTableHeaderRow: true,
			  maxDailiesToShow: 3,
			  dailyExtras: {
				precipitation: true,
				sunrise: false,
				sunset: false,
				wind: false,
				barometricPressure: false,
				humidity: false,
				dewPoint: false,
				uvIndex: false
			  },

			}
		},
		{
            module: "MMM-RemoteCompliments",
            position: "top_center",
            header: "MMM-RemoteCompliments",
            config: {
				fadeSpeed: 4000,
				updateInterval: 10000,
				imageMaxWidth: '250px',
				imageMaxHeight: '250px'
					}
		},
		 {
			module: 'MMM-Cursor',
			config: {
				timeout: 1000
			}
		},
	]
};

/*************** DO NOT EDIT THE LINE BELOW ***************/
if (typeof module !== "undefined") {module.exports = config;}`

![page 2 googlecalendar](https://user-images.githubusercontent.com/106486827/187289068-a78a174e-1079-43c6-94e7-584742da7d87.jpeg)
![Page 1 CX3](https://user-images.githubusercontent.com/106486827/187289139-7d32b742-6f2c-483d-951b-198ec3e6af4a.jpeg)

Let me know if you need something different to troubleshoot.

**Reply by eouia** [reply]
If possible, to make things simple, disable all other modules except googlecalendar and cx3 (and clock)
Then, execute MM on dev mode(stop pm2 and execute ‘npm run start:dev’ on MM dirctory) that will show bunch of log of MM front on the screen(side of mm screen) You can get some information (and errors) about the modules and MM. if you find something suspicious, tell me.

**Reply by CurlyQ12391** [reply]
I simplified the Config down to the basics you mentioned, also thought that having both calendars on the same page would help but they still aren't syncing together. 

I didn't see anything suspicious in the developer tool, but in the terminal I did have a "failed to retrieve device information" pop up...I am not sure if they are related but I attached an image for reference here.
![startdev output](https://user-images.githubusercontent.com/106486827/187314529-cb864e88-cf74-4868-ae7d-c8efff644598.jpeg)

Do you think that perhaps I need to include something in the Config for MMM-CalendarExt3 that would point to MMM-GoogleCalendar? I am not sure what that would be, but wanted to ask since I could have easily forgotten to do that myself.

`{
module: "MMM-GoogleCalendar",
  header: "Skaggs Family Calendar",
  position: "bottom_right",
	config: {
		calendars: [
			{
			symbol: "calendar-week",
			calendarID: "family0xxxxxxx@group.calendar.google.com"
			},
				],
			}
},
{
module: "MMM-CalendarExt3",
	position: "middle_center",
	title: "",
	config: {
		mode: "week",
		weekIndex: -1, // Which week from this week starts in a view. Ignored on mode 'month' 
		weeksInView: 2, //  How many weeks will be displayed. Ignored on mode 'month'
		instanceId: null,
		firstDayOfWeek: 1, // 0: Sunday, 1: Monday
		minimalDaysOfNewYear: 4, // When the first week of new year starts in your country.
			locale: "en-US", 
			maxEventLines: 10,
			firstDayOfWeek: 1,
			useSymbol: true,
			calendarSet: null,
		}
},`

**Reply by eouia** [reply]
Use ‘calendarSet:[],’
I cannot look inside deeper atm. I’ll check it morning.

**Reply by eouia** [reply]
Ok. I'm back

I have looked inside the modules and fixed some minor bugs, but nothing special issue there.



You need to convert the data of `MMM-GoogleCalendar` events as the original default calendar format. But in your configuration, it is not shown. (I wrote it on #22), but you've missed it.

<img width="593" alt="image" src="https://user-images.githubusercontent.com/1720610/187474298-50bb6d6d-046d-4220-93f1-a868792bb38e.png">

**Reply by CurlyQ12391** [reply]
That is absolutely my fault for not understanding where to place the eventTransformer. 

I added that into my Config; however, when I run the config:check I get this error:
    _[30.08.2022 15:40.55.684] [INFO]  Checking file...  /home/pi/MagicMirror/config/config.js
    [30.08.2022 15:40.55.744] [ERROR] Your configuration file contains syntax errors :(
    [30.08.2022 15:40.55.746] [ERROR] Line 117 column 38: Parsing error: Unexpected token ._

Which seems to be the period between **start?** and **date.** Can you see if I did something wrong when pasting in that eventTransformer bit?


`{

		module: "MMM-CalendarExt3",
			position: "middle_center",
			title: "",
			 config: {
				mode: "week",
				weekIndex: -1, // Which week from this week starts in a view. Ignored on mode 'month' 
				weeksInView: 2, //  How many weeks will be displayed. Ignored on mode 'month'
				instanceId: null,
				firstDayOfWeek: 1, // 0: Sunday, 1: Monday
				minimalDaysOfNewYear: 4, // When the first week of new year starts in your country.
				locale: "en-US", 
				maxEventLines: 10,
				firstDayOfWeek: 1,
				useSymbol: true,
				calendarSet: null,
				eventTransformer: (e) => {
				  e.startDate = new Date(e.start?.date || e.start?.dateTime).valueOf()
				  e.endDate = new Date(e.end?.date || e.end?.dateTime).valueOf()
				  e.title = e.summary
				  e.fulldayEvent = (e.start?.date) ? true : false
				  return e
				}
			}
		}`

**Comment by eouia** [comment]
I updated the module to get compatibility with `randomBrainstormer/MMM-GoogleCalendar`. See #22.
